### DISCUSSION TITLE
error: Justfile does not contain recipes when pre-committing

### DISCUSSION THREAD

**Discussion by AlessandroMiola** [casey/just#1984] [Q&A] [answered]
I'm referencing `just` commands in custom `pre-commit` hooks. For some reason, `just` is interpreting *staged files* as *recipes*, which leads to

    trim trailing whitespace.................................................Passed
    fix end of files.........................................................Passed
    check yaml...............................................................Passed
    check toml...............................................................Passed
    ruff (linter)............................................................Failed
    - hook id: ruff-lint
    - exit code: 1

    error: Justfile does not contain recipes 
    `folder_1/src/alembic/versions/fb4554c0e46f_first_commit.py`, 
    `folder_1/src/alembic/env.py`, `folder_1/src/crud/user.py`, or `config.py`.

    ruff (formatter).........................................................Failed
    - hook id: ruff-format
    - exit code: 1

    error: Justfile does not contain recipes 
    `folder_1/src/alembic/versions/fb4554c0e46f_first_commit.py`, 
    `folder_1/src/alembic/env.py`, `folder_1/src/crud/user.py`, or `config.py`.

My project tree is as follows (at least to what pertains to the files listed within the error message):

    |-- folder_1
        |-- src
            |-- alembic
            |   |-- env.py
            |   |-- versions
            |       |-- fb4554c0e46f_first_commit.py
            |-- crud
            |   |-- user.py
            ...
    |-- folder_2
        |-- src
        ...
    .pre-commit-config.yaml
    config.py
    justfile
    poetry.lock
    pyproject.toml

My `.pre-commit-config.yaml` is as follows:

    default_language_version:
      python: python3.10

    repos:
      - repo: https://github.com/pre-commit/pre-commit-hooks
        rev: v4.4.0
        hooks:
          - id: trailing-whitespace
          - id: end-of-file-fixer
          - id: check-yaml
          - id: check-toml
      - repo: local
        hooks:
          - id: ruff-lint
            name: ruff (linter)
            entry: just ruff-fix
            language: system
            types: [python]
          - id: ruff-format
            name: ruff (formatter)
            entry: just ruff-format
            language: system
            types: [python]

My `justfile` is as follows:

    help:
      @just --list

    install:
      @echo "🚀 Installing dependencies"
      @poetry install --with dev

    install-pre-commit:
      @echo "🚀 Setting up the hooks"
      @poetry run pre-commit install

    check-project:
      @echo "🚀 Checking consistency between poetry.lock and pyproject.toml"
      @poetry check --lock
      @echo "🚀 Running the hooks against all files"
      @poetry run pre-commit run --all-files

    ruff:
      @echo "🚀 Linting the project with Ruff"
      @poetry run ruff check .

    ruff-show-violations:
      @echo "🚀 Linting the project with Ruff and show violations"
      @poetry run ruff check --output-format="grouped" .

    ruff-fix:
      @echo "🚀 Linting the project with Ruff and autofix violations (where possible)"
      @poetry run ruff check --fix .

    ruff-format:
      @echo "🚀 Formatting the code with Ruff"
      @poetry run ruff format .

    ruff-format-check:
      @echo "🚀 Listing files Ruff would reformat"
      @poetry run ruff format --check .

    lint-and-format: ruff-fix ruff-format

Versions:

    $ just --version
    just 1.23.0

    $ pre-commit --version
    pre-commit 3.7.0

Ofc, substituting [`poetry run ruff check --fix .`] to [`just ruff-fix`] and [`poetry run ruff format .`] to [`just ruff-format`] into `.pre-commit-config.yaml` just :) works, but I'd like to use `just` commands.

Can someone help? What am I doing wrong? A similar setup used to work elsewhere for me.

**Comment by cmb84scd** [comment, marked as answer]
I came across this question while trying to solve the same problem and wanted to share something I have just discovered.

If I add `pass_filenames: false` to the pre-commit hook and use a just recipe, it then works as expected. I found this [article](https://medium.com/@fistralpro/pytest-pre-commit-hook-b492edd0560e) semi-helpful in understanding what was going on and although I still don't fully get it, it would appear pre-commit passing in the files is causing conflicts with Just.

I don't know if this is the best/right solution but I would be interested to know what others think and why it's necessary to add it to get pre-commit to work nicely with just. Thanks.

**Reply by muzimuzhi** [reply]
> If I add `pass_filenames: false` to the pre-commit hook and use a just recipe, it then works as expected.

Thank you. It worked for me!

Here the link to [`pass_filenames`](https://pre-commit.com/#hooks-pass_filenames) doc.
